---
title: "Data Types in R"
subtitle: "BODL R Training"
author: "Dr Stevie Pederson"
institute: "Black Ochre Data Labs<br>Telethon Kids Institute"
date: "2023-02-17"
date-format: long
title-slide-attributes:
    data-background-color: "#3c3c44"
    data-background-image: assets/bodl_logo_white_background.jpg
    data-background-opacity: "0.3"
    data-background-size: "90%"
editor: source
format: 
  revealjs:
    theme: [../custom.scss]
    code-line-numbers: false
    width: 1024
    height: 768
    sansfont: Times New Roman
    logo: assets/bodl_logo_white_background.jpg
---

```{r setup, echo = FALSE, include = FALSE}
if (interactive()) setwd(here::here("docs"))
knitr::opts_chunk$set(
  echo = TRUE, include = TRUE, warning = FALSE, message = FALSE, 
  fig.align = "center", results = 'hide', fig.show = "asis",
  fig.width = 6
)
```

## Welcome To Day 2

*I acknowledge & pay my respects to the Kaurna people as traditional owners & custodians of this land*

## Welcome To Day 2

Today's Topics:

1. __`R` Objects & Data Types__
2. Text Manipulation
3. Basic Statistical Analysis
<!-- 4. Writing Functions and Iteration -->
<!-- 5. The Bioconductor Project -->


# `R` Objects {background-color="#3c3c44"}

## Before we start {.build}

1. Clear your `Global Environment`  (![](assets/broom.png))
    + `Session` > `Restart R`
2. Create a new R Script:  
    + `File > New File > R Script`
    + Save as `DataTypes.R`

## Recap Of Day 1

We learned how to:

1. Load a csv using `readr::read_csv()`
2. Look through & summarise a `data.frame`
3. Pipe data objects (`%>%` or `|>`)
4. Generate plots with `ggplot2`
<!-- 5. Use Rmarkdown -->


## `R` Objects

- Main data type so far has been a `data.frame` 
    - `tibble` = `data.frame` with nice wrapping paper
- Each column has the same type of data, e.g. `numeric`, `character` etc.
- The columns can be different types to each other
- In `R` each column is a `vector`

# Vectors {background-color="#3c3c44"}

## Vectors

The key building blocks for `R` objects: **Vectors**

- There is no such thing as a *scalar* in `R`
    + A scalar is a single number
- Everything is based around the concept of a *vector*

**What is a vector?**

. . .

### Definition

A vector is zero or more values of the **same type**

## Examples

A simple vector would be

```{r show-seq, echo=FALSE, include=TRUE, results='markup'}
1:10
```

__What type of values are in this vector?__

## Examples

Another vector might be

```{r show-chars, echo=FALSE, include=TRUE, results='markup'}
c("a", "cat", "video")
```

__What type of values are in this vector?__

## Examples

__What type of values are in this vector?__

```{r show-more-cars, echo=FALSE, include=TRUE, results='markup'}
c("742", "Evergreen", "Tce")
```


## The 4 Atomic Vector Types

- *Atomic Vectors* are the building blocks for everything in `R`
- There are six atomic vector types
    + Four we'll focus on
    - Two we can ignore
    
. . .

1. _logical_
2. _integer_
3. _numeric_
4. _character_

. . .

(We'll ignore `complex` and `raw` vectors)


## Logical Vectors

1. **logical**: Can only hold the values `TRUE` or `FALSE`

```{r logi-vec}
logi_vec <- c(TRUE, TRUE, FALSE)
print(logi_vec)
```


## Integer Vectors

1. logical
2. **integer**: Counts, ranks or indexing positions 

```{r int-vec}
int_vec <- 1:5
print(int_vec)
```


## Double (i.e. Double Precision) Vectors

1. logical
2. integer
3. **double**: Usually (& *lazily*) referred to as `numeric`

```{r bdl-vec, results='hide'}
dbl_vec <- c(0.618, 1, 2.618)
print(dbl_vec)
```

**Why are these called doubles?**

## Character Vectors

1. logical
2. integer
3. double
4. **character**

```{r char-vec, include=TRUE, results='hide'}
char_vec <- c("blue", "red", "green")
print(char_vec)
```

## The 4 Atomic Vector Types

These are the basic building blocks for all `R` objects

1. logical
2. integer
3. double
4. character

. . .

**All `R` data structures are built on these!**

## Properties of a vector

What *four* defining properties might a vector have?

::: {.incremental}

1. The *values*
2. *Length*
    + Accessed by the function `length()`
3. *Type*
    + Accessed by the function `typeof()`
    + Preferable to `class()` for vectors
4. Any optional & additional *attributes* 
    + Accessed using `attributes()`
    + Holds data such as `names` etc.
    
:::

## Properties of a vector

__Let's try them on our vectors__

```{r try-funs, results='hide'}
typeof(char_vec)
length(int_vec)
attributes(logi_vec)
class(dbl_vec)
typeof(dbl_vec)
```

__Were you surprised by any of the results?__

## Working with Vectors

We can *combine* two vectors in `R`, using the function `c()`

```{r show-c}
c(1, 2)
```

- The numbers `1` & `2` were both vectors with `length()` == 1

- We have combined two vectors of length 1, to make a vector of length 2

. . .

- A vector of length zero is able to be formed using `c()`
    + This also represents the value `NULL` in `R`

## Working with Vectors 

__What would happen if we combined two vectors of different types?__

. . .

Let's try & see what happens:

```{r new-vec, results='hide'}
new_vec <- c(logi_vec, int_vec)
print(new_vec)
typeof(new_vec)
```

. . .

__What happened to the `logical` values?__

. . .

`R` will coerce them into a common type (i.e. integers).

# Coercion {background-color="#3c3c44"}

## Coercion

::: {.incremental}

- If combining values of the same type $\implies$ no coercion
- If combining values of different types
    + `R` will coerce to the simplest possible value
    + ` `1) `logical`, 2) `integer`, 3) `numeric` and 4) `character`
    + These are ranked by memory, i.e. `logical` values occupy a *single byte*
- When no coercion is possible `NA` values will be produced
- We can explicitly coerce values using `as.*()`
    + `as.logical()`, `as.integer()`, `as.numeric()`, `as.character()`
    + `as.*()` functions are also implemented for other data-types

:::


## Discussion Questions

::: {.incremental}

1. **What other types could `logical` vectors be coerced into?**
    + Try coercing `logi_vec` to the three other types
2. **Can `numeric` vectors be coerced into `logical` vectors?**
    + Let's try coercing `dbl_vec` to `logical`
3. **Can `numeric` vectors be coerced to `integer` vectors?**
    + Is this what you expected?
    + Anything after the decimal point is simply dropped
3. **Can `character` vectors be coerced into `numeric` vectors?**

:::

. . . 

```{r simp-vec}
simp_vec <- c("742", "Evergreen", "Terrace")
as.numeric(simp_vec)
```

## `NA` and Special Values

::: {.incremental}

- When coercing to more complex types $\implies$ almost never `NA`
- When coercing to simpler data types $\implies$ `NA` values are common
- `NA` values are typed in `R`
    + Type `NA_` in the Console and see what options are shown
    + Can cause problems with code if unexpected
- `NaN` is very different to `NA`
    + This means *Not-a-Number* (e.g. `0/0`)
- `Inf` is a `numeric` value
- `NULL` is an empty value with type `NULL`
    
:::

# Subsetting Vectors {background-color="#3c3c44"}

## Subsetting Vectors

#### Single Brackets

One or more elements of a vector can be called using `[]`

```{r single-brackets}
new_vec
new_vec[2]
new_vec[1:3]
```

This returns an object of the **exact same type**, but with only the elements specified
<br><br>

:::: {.fragment}

#### Double Brackets

Double brackets (`[[]]`) can **only** be used to return __single elements__

```{r double-brackets}
new_vec[[2]]
```

::: {.incremental}

- If you try `new_vec[[1:3]]` you'll receive an error message
- Only a vector of length `1` can be placed here

:::

::::

## Subsetting Vectors

If a vector has name attributes, we can call values by name

- The inbuilt vector `euro` contains conversion rates at the time of Euro launch

```{r eruo-single-brackets}
head(euro)
attributes(euro)
names(euro)
euro["ESP"]
```

## Subsetting Vectors

*Try repeating the* call-by-name *approach using double brackets*

```{r euro-double-brackets}
euro["ESP"]
euro[["ESP"]]
```
<br>

. . .

__What was the difference in the output?__

1. Using `[]` returned a subset of the vector with all other `attributes`
2. Using `[[]]` removed the vector-level `attributes` & just gave the value

## Extracting Multiple Values 

**What is really happening in this line?**

```{r euro-by-position}
euro[1:3]
```

. . .

- `1:3` is an `integer` vector
- We are using an `integer` vector to call values by position

## Discussion Questions

__Would it better to call values by position, or by name?__

Things to consider:

- Which is easier to type "on the fly"?
- Which is easier to read?
- Which is more robust to undocumented changes in an object?


## Extracting Multiple Values 

We can actually define the vectors we use to subset values

```{r euro-by-int-vec}
int_vec
euro[int_vec]
```

. . .

- This can make code harder to read
- Extremely useful when regularly calling values by name

## Vector Operations

`R` Functions are designed to work on vectors

```{r dbl-vec-logical-test, results='hide'}
dbl_vec > 1
```

This is one of the real strengths of `R`

## Vector Operations

We can also combine the above logical test and subsetting

```{r subset-dbl-vec-by-logical}
dbl_vec
dbl_vec > 1
dbl_vec[dbl_vec > 1]
```

. . .

- Here we've used a logical vector to extract values 
    + Must be the same length as the main vector
    + Usually used to self-subset

## Vector Operations

An additional logical test: `%in%` (read as: "*is in*")

```{r print-dbl-int-vecs}
dbl_vec
int_vec
```

We can use this to check of values from one vector are in another

. . .

```{r dbl-is-in-int, results='hide'}
dbl_vec %in% int_vec
```

. . .

Returns `TRUE/FALSE` for each value in `dbl_vec` if it **is in** `int_vec`

NB: `int_vec` was coerced silently to a `double` vector

. . .

```{r subset-dbl-by-is-in-int}
dbl_vec[dbl_vec %in% int_vec]
```

## Vectors

Any questions...?

# Matrices {background-color="#3c3c44"}

## Matrices

- Vectors are strictly one dimensional and have a `length` attribute.
- A `matrix` is the **two dimensional** equivalent
- Very common in Bioinformatics
    + Same measurement (e.g. gene expression) taken across all samples

::: {columns}
::: {.column width="40%"}

![Example count matrix^[Image taken from https://bioconnector.github.io/workshops/r-rnaseq-airway.html]](assets/count_matrix.png){width="300" height="265"}

:::
:::

::: {.notes}
This is a classic structure for gene expression:
1. Each row would be the same gene
2. Each column would be the same sample
:::

## Matrices

- Matrices can only hold __one__ type of value
    + i.e. logical, integer, double, character
- Have additional attributes such as `dim()`, `nrow()` `ncol()`
- Can have __optional__ `rownames()` & `colnames()`
- [Very useful]{.underline} for linear algebra

. . .

```{r, results='hide'}
int_mat <- matrix(1:6, ncol=2)
int_mat
```


## Matrices

__Some commands to try:__

```{r, results='hide'}
dim(int_mat)
typeof(int_mat)
class(int_mat)
attributes(int_mat)
colnames(int_mat)
length(int_mat)
```

__Ask questions if anything is confusing__

## Matrices

- Use square brackets to extract values by row & column
- The form is `x[row, col]`
- Leaving either `row` or `col` blank selects the entire row/column

```{r, results='hide'}
int_mat[2, 2]
int_mat[1,]
```

__How would we just get the first column?__
<br><br>

. . .

_NB: Forgetting the comma will treat the matrix as a vector_ **running down the columns**

```{r}
int_mat
int_mat[5]
```

## Matrices

**Requesting a row or column that doesn't exist is the source of a very common error message**

```{r, results='markup'}
#| output-location: default
dim(int_mat)
```

. . .

Given we have `r nrow(int_mat)` rows, we *should* expect an error if we ask for a row > `r nrow(int_mat)`

```{r, eval=TRUE, error=TRUE}
int_mat[5,]
```

## Matrices

- We can access column names or row names
    + `colnames()` and `rownames()`
- We can also set them

```{r show-colnames, results='markup'}
colnames(int_mat)
```

To set them we use the `colnames()<-` strategy

```{r set-colnames}
colnames(int_mat) <- c("A", "B")  
colnames(int_mat)
```

. . .

Both column and row names can be returned using `dimnames()`

```{r mat-dimnames}
dimnames(int_mat)
```


## Arrays

- Arrays extend matrices to 3 or more dimensions
- Beyond the scope of today, but we just have more commas in the square brackets
- Less commonly used in bioinformatics

```{r iris, results='markup'}
dim(iris3)
length(iris3)
dimnames(iris3)
```

## Summary

- These are the basic *homogeneous* data types of `R`
    + 1D: Vectors
    + 2D: Matrices
    + >2D: Arrays

. . .

- All are essentially just vectors

# Heterogeneous Data Types {background-color="#3c3c44"}

## Heterogeneous Data Types

::: {.columns}
::: {.column width="50%}

Dimension   | Homogeneous | [Heterogeneous]{.fragment fragment-index=1}
------------|-------------|--------------
1d          | `vector`    | [`list`]{.fragment fragment-index=2}
2d          | `matrix`    | [`data.frame`]{.fragment fragment-index=3}
3d+         | `array`     |

Table: *Main data types in R*

:::
:::

[We've already seen `data.frame` objects...]{.fragment fragment-index=4}

## List Objects

A `list` is a __heterogeneous__ vector.

::: {.incremental}

- Each element is an `R` object
    + Could be a `vector` or `matrix`
    + Could be another `list`
    + Any other `R` object type we haven't seen yet

:::

::: {.fragment}

**These are incredibly common in `R`**

:::

## Lists

We can put all our vectors into a single list

```{r}
my_list <- list(
  logical = logi_vec,
  integer = int_vec,
  doubles = dbl_vec,
  characters = char_vec
)
my_list
```

. . .

- This is a list of length `r length(my_list)`
- At each position is another `R` object
- No way to force row or columns on this

. . .

- We did add names though (note the `$` in the main object)

```{r}
names(my_list)
```

## Lists

We can call the individual components of a list using the `$` symbol followed by the name

```{r, results='hide'}
my_list$logical
my_list$integer
```

*Note that each component is quite different to the others.*

## Lists

Many `R` functions provide output as a list

```{r, results='hide'}
t_results <- t.test(dbl_vec)
class(t_results)
typeof(t_results)
t_results
```

NB: There is a function (`print.htest()`) that tells `R` how to print the results to the Console

## Lists

Explore the various attributes of the object `t_results`

```{r, results='hide'}
str(t_results)
attributes(t_results)
length(t_results)
```

. . .

We can call the individual components of a list using the `$` symbol followed by the name

```{r, results='hide'}
t_results$statistic
t_results$conf.int
t_results$method
```


## Subsetting Lists

A `list` is a `vector` so we can also subset using the `[]` method

```{r, results='hide'}
t_results[1]
typeof(t_results[1])
```

- Using single square brackets returns a `list` 
    - i.e. an object which is a subset of the larger object, but of the *same type*

. . .

- Double brackets again retrieve a single element of the vector
    + Returns the actual `R` object at that position

```{r, results='hide'}
t_results[[1]]
typeof(t_results[[1]])
```

. . .

*When would we use either method?*

## Subsetting Lists

We can also use names instead of positions

```{r, results='hide'}
t_results[c("statistic", "p.value")]
t_results[["statistic"]]
t_results$statistic
```

. . .

- `t_results[["statistic"]]` and `t_results$statistic` produce **identical** output
    + `$` allows auto-completion in RStudio

## Lists

- Note also the **Environment** Tab in the top right of RStudio
- Click the arrow next to `t_results` to expand the entry
- This is the output of `str(t_results)`

# Data Frame Objects {background-color="#3c3c44"}

## Data Frames

(Finally!)

- These are the *most common type of data* you will work with

. . .

- Each column is a `vector`
    + i.e. will be coerced to the same type
    + Columns can be different vector types 
    + Column vectors **MUST** be the same length
    
. . .

- A relationship across rows is also expected


## Data Frames

- Surface similarity to matrices, but are **specifically for heterogeneous data**
    - Not useful at all for linear algebra
    - Brilliant for presenting bioinformatics result tables
. . .

- Have many of the same attributes as matrices
    + `dim()`, `nrow()`, `ncol()`, `rownames()`, `colnames()`
    
. . .

- `colnames()` & `rownames()` are **NOT** optional & are assigned by default
    - No `dimnames()` attribute
    


## Data Frames

Let's load `toothData` again

```{r, eval=TRUE, results='hide'}
library(tidyverse)
toothData <- read_csv("data/toothData.csv")
```

__Try these commands__

```{r}
glimpse(toothData)
colnames(toothData)
dim(toothData)
nrow(toothData)
rownames(toothData)
```

. . .

- Note that rownames were automatically assigned
- This is a character vector!!! [`"1"` $\neq$ `1`]{.fragment}
    + For `tibble` objects they're never printed to screen 
    + For `tibble` objects they always take these values


## Data Frames

Individual entries can also be extracted using the square brackets

```{r, results='hide'}
toothData[1:2, 1]
```

. . .

We can also refer to columns by name (same as matrices)

```{r, results='hide'}
toothData[1:2, "len"]
```

. . .

The `dplyr` (i.e. `tidy`) way of doing this would be:

```{r}
toothData |>
  slice(1:2) |>
  select(len)
```


## Data Frames

Remember that each column is a separate vector

- We can call each column vector of a `data.frame` using the `$` operator
    + Very different results to `toothData[1:2, "len"]`

```{r, results='hide'}
toothData$len[1:2]
```

. . .

**No `$` equivalent exists for rows!!!**

## Data Frames

- `R` is *column major* by default (as is `FORTRAN` & Matlab)
- Many other languages are *row major*, e.g. C/C++, Python
- `R` was designed for statistical analysis, but has developed capabilities **far beyond this**

(We will see this advantage this afternoon)

# Data Frames & Lists {background-color="#3c3c44"}

## Data Frames & Lists

__Data frames are actually special cases of lists__


```{r}
is.data.frame(toothData)
is.list(toothData)
```

. . .

- Each column of a `data.frame` is a component of a `list`
- The components must all be vectors of the **same length**
- Data Frames can be treated identically to a `list`
- Have additional subsetting operations and attributes
    + e.g. Lists cannot be selected by row

## Data Frames & Lists

Forgetting the comma, now gives a completely different result to a matrix!

```{r, eval=FALSE}
toothData[1]
```

__Was that what you expected?__

[Hopefully this is now intuitive knowing it's really a list]{.fragment}

<br>

__Try using the double bracket method__

## More Errors

**What do you think will happen if we type:**

```{r, error=TRUE}
#| output-location: fragment
toothData[5]
```


# Working With R Objects {background-color="#3c3c44"}

## Name Attributes

- Each different data type can have different types of names

| Object | `names()` | `rownames()` | `colnames()` | `dimnames()` |
|:------ |:--------- |:------------ |:------------ |:------------ |
| Vector | Optional  | &#10060;     | &#10060;     | &#10060;     |
| List   | Optional  | &#10060;     | &#10060;     | &#10060;     |
| Matrix | &#10060;  | Optional     | Optional     | Optional     |
| Data Frame | &#9989; |  &#9989;   |  &#9989;     | &#9989;      |

. . .


- `names()` for a matrix can be added but are without relevance
- `rownames()` for a `tibble` are ignored
- `dimnames()` for a `data.frame` are rarely accessed like this
- `names()` = `colnames()` for a `data.frame`


## Name Attributes

__How do we assign names?__

1. We can do so when creating an object

```{r}
named_vec <- c(a = 1, b = 2, c = 3)
```

. . .

2. We can add names to an existing vector

```{r}
names(int_vec) <- c("a", "b", "c", "d", "e")
```

## Removing Name Attributes

__Can we remove names?__

. . .

The `NULL`, or empty vector, in `R` is created using `c()`

```{r, results='hide'}
null_vec <- c()
length(null_vec)
```

. . .

We can use this to remove names

```{r}
names(named_vec) <- c()
```

. . .

*Don't forget to put the names back...*

```{r}
names(named_vec) < c("a", "b", "c")
```


## Matrices 

We can assign row names & column names to matrices after creation

```{r}
colnames(int_mat) <- c("low", "high")
```

Or using `dimnames()`

```{r, results='hide'}
dimnames(int_mat) <- list(
  c("first", "second", "third"),
  c("low", "high")
)
```

This a list of `length` 2 with `rownames` then `colnames` as the components.

**Why were the elements not named `rownames` or `colnames`?**

<!-- ## A word of caution -->

<!-- ```{r, warning=TRUE} -->
<!-- test_mat <- matrix(int_vec, ncol = 2) -->
<!-- ``` -->

<!-- __Did we successfully form an `R` object?__ -->

<!-- __What has happened here?__ -->

<!-- __This is a major (and warranted) criticism made of R__ -->

## Lists

Names can be accessed and set in a similar way for lists

```{r}
names(my_list)
```

. . .

We can easily overwrite these

```{r}
names(my_list) <- c("a", "b", "c", "d")
names(my_list)
```

## Lists

__What happens if we try this?__

```{r}
my_list$logical <- logi_vec
```

. . .

We can easily add elements to a list using this strategy

. . .

Or we can remove them from a list

```{r}
my_list$logical <- NULL
```


<!-- ## Lists -->
<!-- We can coerce vectors to lists as well -->
<!-- ```{r} -->
<!-- int_list <- as.list(named_vec) -->
<!-- ``` -->

## Data Frames

This is exactly the same as creating lists, but

__The `names` attribute will also be the `colnames()`__

```{r}
my_df <- data.frame(doubles = dbl_vec, logical = logi_vec)
names(my_df) == colnames(my_df)
```

. . .

We can also add columns as for lists

```{r}
my_df$letters <- c("a", "b", "c")
```

. . .

Up to us to ensure this is sensible

. . .

Can remove columns the same way as lists

```{r}
my_df$letters <- NULL
```

Or use `dplyr::select()`


## Data Frames

__What happens if we try to add components that aren't the same length?__

```{r, error=TRUE}
#| output-location: fragment
my_df <- data.frame(
  integers = int_vec, 
  doubles = dbl_vec,
  logical = logi_vec
)
```


# Morning Tea


<div class="footer" style="text-align:center;width:25%">
[Home](http://uofabioinformaticshub.github.io/RAdelaide-July-2018/)
</div>
